### PR TITLE
Set read timeout on file descriptor

### PR DESCRIPTION
**Pull request by corny**
Using `poll` before `read` is simple and works with timeouts.

### CODE REVIEW

**Comment by mdlayher** [comment]
This looks promising! Will do a review tomorrow. Excited to get some proper timeouts in here. I've been meaning to do it for Linux for ages.

**Comment by corny** [comment]
It doesn't seem to work properly yet. I'm still debugging.

**Comment by corny** [comment]
I did `int(duration.Nanoseconds()/1000)` instead of `int(duration.Nanoseconds()/1000/1000)`. The timeout works now.

I'm not sure which errors to return. Since `poll` also exists in Linux this can (better) generic solution instead of using of reading from non-blocking sockets periodically.

**Comment by corny** [comment]
I figured out how to set read timeouts on file descriptors. This might be the best solution. Please check out my changes and do some tests.

File: raw_bsd.go [hunk 1]
```diff
@@ -111,13 +117,38 @@ func listenPacket(ifi *net.Interface, proto Protocol) (*packetConn, error) {
...
+	var hasTimeout bool
+
+	p.timeoutMu.Lock()
+	if !p.rtimeout.IsZero() {
+		timeout = p.rtimeout.Sub(time.Now())
+		hasTimeout = true
+	}
+	p.timeoutMu.Unlock()
+
+	if hasTimeout && timeout < time.Microsecond {
```
Comment (mdlayher) [inline raw_bsd.go]: Is this check necessary? If someone does something like `time.Duration(1)` (which is 1ns) it's usually a programmer error.

File: raw_bsd.go [hunk 2]
```diff
@@ -111,13 +117,38 @@ func listenPacket(ifi *net.Interface, proto Protocol) (*packetConn, error) {
...
+		return 0, nil, &timeoutError{}
+	}
+
+	tv := &unix.Timeval{
+		Sec:  int64(timeout / time.Second),
+		Usec: int64(timeout % time.Microsecond),
+	}
+
+	if _, _, err := syscall.Syscall(syscall.SYS_IOCTL, uintptr(p.fd), syscall.BIOCSRTIMEOUT, uintptr(unsafe.Pointer(tv))); err != 0 {
```
Comment (mdlayher) [inline raw_bsd.go]: This seems like an unusual API.  Any reason you didn't stick with poll or something?

File: raw_linux.go [hunk 3]
```diff
@@ -97,67 +91,55 @@ func newPacketConn(ifi *net.Interface, s socket, pbe uint16, sleeper sleeper) (*
...
-				Addr: nil,
-				Err:  &timeoutError{},
-			}
-		default:
-			// Not timed out, keep trying
-		}
 
 		// Attempt to receive on socket
 		n, addr, err = p.s.Recvfrom(b, 0)
 		if err != nil {
```
Comment (mdlayher) [inline raw_linux.go]: On Linux at least, it looks like we need to check for `EAGAIN` here and return a `&timeoutError{}` too.

File: raw_bsd.go (hunk 2 above)
Comment (corny) [reply raw_bsd.go]: `poll` is more complex than just setting a read timeout. `poll` is helpful if you have multiple file descriptors and/or multiple events you want to be notified about. There is also a `SetBpfTimeout` function in the standard library, but these BPF functions are going to be removed. In my point of view setting a timeout by using `BIOCSRTIMEOUT` is the simplest way. It's documented here:
https://www.freebsd.org/cgi/man.cgi?query=bpf&sektion=4

File: raw_linux.go (hunk 3 above)
Comment (corny) [reply raw_linux.go]: The man pages say that `EAGAIN` only happens for non-blocking sockets. We have a blocking socket here.

File: raw_bsd.go (hunk 1 above)
Comment (mdlayher) [reply raw_bsd.go]: Cool, can you leave a comment to that effect?

File: raw_linux.go (hunk 3 above)
Comment (mdlayher) [reply raw_linux.go]: Interesting, when running this locally on Linux, I'm seeing:

```
$ time sudo ./arpc -d 2s -ip 192.168.1.100
2017/11/02 12:37:23 resource temporarily unavailable
sudo ./arpc -d 2s -ip 192.168.1.100  0.00s user 0.01s system 0% cpu 1.945 total
```

Which is just `EAGAIN` in string form.  I think that the setsockopt call might be setting non-blocking for you.

File: raw_bsd.go (hunk 1 above)
Comment (corny) [reply raw_bsd.go]: done

File: raw_linux.go (hunk 3 above)
Comment (corny) [reply raw_linux.go]: I am going to check this later.

Comment (corny) [reply raw_linux.go]: On non-blocking sockets `EAGAIN` seems to indicate a timeout. I modified my changeset to return a `timeoutError` in that case.

**Comment by corny** [comment]
The `recvfrom` syscall does not return immediately on `close` or `shutdown` of the socket. It only returns after the timeout. We should find a solution for this.

**Comment by mdlayher** [comment]
That approach seems fine to me, so long as we can hide the complexity under our implementation of `net.PacketConn`.

**Comment by mdlayher** [comment]
Thanks for taking this on, by the way. I really appreciate it.

**Comment by mdlayher** [comment]
Fixes #4.

**Comment by corny** [comment]
Tests should be fine now. Please review and merge.

File: raw.go [hunk 4]
```diff
@@ -19,6 +19,10 @@ const (
 
 	// ProtocolWoL specifies the Wake-on-LAN protocol.
 	ProtocolWoL Protocol = 0x0842
+
+	// Maximum read timeout per syscall.
+	// It is required because read/recvfrom won't be interrupted on closing of the file descriptor.
+	readTimeout time.Duration = time.Millisecond * 200
```
Comment (mdlayher) [inline raw.go]: You can omit the `time.Duration` part and it will be inferred from the right side.

**Comment by corny** [comment]
I removed the `time.Duration` type. VS code shows the variable as `int` value. But it does not change anything.

**Comment by mdlayher** [comment]
Thanks again. Will merge once the Mac build passes.

**Comment by corny** [comment]
Finally done ... after 42 minutes :sleeping:

**Comment by mdlayher** [comment]
Yeah, the Travis Mac runner is pretty awful sometimes.

**Comment by corny** [comment]
Thanks a lot!

**Pull request merged**